### PR TITLE
Add MultiLang capability to Winter.Sitemap.

### PR DESCRIPTION
**Pull request by mjauvin**
Related PRs needed for this: https://github.com/wintercms/wn-sitemap-plugin/pull/11

Add MultiLang capability to Winter.Sitemap for Winter.Blog / CMS Pages menuItemTypes.

### CODE REVIEW

File: Plugin.php [hunk 1]
```diff
@@ -366,4 +371,48 @@ public function translateRawPlural($string, $count = 0, $params = [], $locale =
...
+        });
+
+        Event::listen('winter.sitemap.makeUrlSet', function ($definition, $xml, $urlSet) {
+            $botsRegex = '/(Apple|Bing|Google|DuckDuck|Yandex|Exa|Face)[bB]ot|Baiduspider|Slurp|Sogou/';
+            if (preg_match($botsRegex, Request::server('HTTP_USER_AGENT', ''))) {
+                $nsUrl = 'http://www.w3.org/1999/xhtml';
+            } else {
+                // hack to force browser to properly render the XML sitemap
+                $nsUrl = 'xmlns:xhtml-namespace-definition-URL-here';
+            }
```
Comment (matteotrubini) [inline Plugin.php:400]: if this hack is just for debugging purpose, would not be safer to test if `debug` exists in query string, avoiding potential problem with non-listed bots by default?

Comment (mjauvin) [reply Plugin.php]: Fixed.

File: Plugin.php [hunk 2]
```diff
@@ -366,4 +373,56 @@ public function translateRawPlural($string, $count = 0, $params = [], $locale =
...
+                    case 'blog-category':
+                    case 'all-blog-categories':
+                        return Classes\MLBlogCategoryModel::resolveMenuItem($item, $url, $theme);
+                    case 'blog-post':
+                    case 'category-blog-posts':
+                    case 'all-blog-posts':
+                        return Classes\MLBlogPostModel::resolveMenuItem($item, $url, $theme);
+                }
+            }
+            return false;
```
Comment (LukeTowers) [inline Plugin.php:401]:
```suggestion
            switch ($item->type) {
                case 'cms-page':
                    return Classes\MLCmsPage::resolveMenuItem($item, $url, $theme);
                case 'static-page':
                    return Classes\MLStaticPage::resolveMenuItem($item, $url, $theme);
                case 'blog-category':
                case 'all-blog-categories':
                    return Classes\MLBlogCategoryModel::resolveMenuItem($item, $url, $theme);
                case 'blog-post':
                case 'category-blog-posts':
                case 'all-blog-posts':
                    return Classes\MLBlogPostModel::resolveMenuItem($item, $url, $theme);
                default:
                    return false;
            }
```

File: Plugin.php [hunk 3]
```diff
@@ -366,4 +373,56 @@ public function translateRawPlural($string, $count = 0, $params = [], $locale =
...
+
+        Event::listen('winter.sitemap.makeUrlSet', function ($definition, $xml, $urlSet) {
+            if (Config::get('app.debug', false)) {
+                // hack to force browser to properly render the XML sitemap
+                $nsUrl = 'xmlns:xhtml-namespace-definition-URL-here';
+            } else {
+                $nsUrl = 'http://www.w3.org/1999/xhtml';
+            }
+            $urlSet->setAttribute('xmlns:xhtml', $nsUrl);
+        });
```
Comment (LukeTowers) [inline Plugin.php:412]: This should just be done by the Winter.Sitemap plugin itself.

**Review by LukeTowers — CHANGES_REQUESTED** [review]

File: Plugin.php (hunk 2 above)
Comment (mjauvin) [reply Plugin.php:401]: This was done on purpose for when those plugins are NOT installed/enabled.

File: Plugin.php (hunk 3 above)
Comment (mjauvin) [reply Plugin.php:412]: Why? This is only needed if xhtml:link nodes are addded (which is not the case in the base plugin)

Comment (LukeTowers) [reply Plugin.php:412]: Do you have a link to an article or something that covers the reason for this hack?

File: Plugin.php (hunk 2 above)
Comment (LukeTowers) [reply Plugin.php:401]: Sure, but if they're not installed / enabled then they won't have entries in the sitemap.

File: Plugin.php (hunk 3 above)
Comment (mjauvin) [reply Plugin.php:412]: No, it's just that when there are `xhtml:link` nodes, you need to add the namespace declaration for them, but chrome has problem formatting the xml so it ends up like this:

Screenshot from 2023-02-28 21-34-24

instead of like this (with the hack):

image

File: Plugin.php (hunk 2 above)
Comment (mjauvin) [reply Plugin.php:401]: Are you playing devil's advocate here? lol

There may be cases where left-over menuItemTypes from the blog plugin are in the Sitemap, and the blog plugin has been disabled/uninstalled without getting rid of these items... in which case the sitemap would silently (since this can easily be overlooked) break. With these conditional blocks in place, the sitemap would keep working.

File: Plugin.php (hunk 3 above)
Comment (AIC-BV) [reply Plugin.php:412]: Can confirm that the hack does what it needs to do.
Readable for us, and readable for bots (I used SiteGuru and it can succesfuly read my sitemap)

File: Plugin.php (hunk 1 above)
Comment (AIC-BV) [reply Plugin.php]: Just some thinking;
How can the bots read my sitemap on staging environment? Debug mode will be active

I don't know how these work; but things like SiteGuru and ScreamingFrog should still work on staging environments?

**Comment by AIC-BV** [comment]
I added these 5 file changes and 1 sitemap Definition.php change,

CMS pages work out of the box
Static pages aren't working
I don't have the blog plugin.

I will now try to add my own type

**Comment by AIC-BV** [comment]
Also wondering if these are still required? Or if they could be merged in any way. They are rather similar!
image

File: Plugin.php [hunk 4]
```diff
@@ -366,4 +373,56 @@ public function translateRawPlural($string, $count = 0, $params = [], $locale =
...
+
+    public function extendWinterSitemap()
+    {
+        Event::listen('winter.sitemap.processMenuItems', function ($item, $url, $theme, $apiResult) {
+            if ($item->type === 'cms-page') {
+                return Classes\MLCmsPage::resolveMenuItem($item, $url, $theme);
+            }
+
+            if (PluginManager::instance()->exists('Winter.Pages')) {
+                if ($item->type === 'static-page') {
```
Comment (AIC-BV) [inline Plugin.php:385]: if ($item->type === 'static-page' || $item->type === 'all-static-pages') {

**Comment by mjauvin** [comment]
> Also wondering if these are still required? Or if they could be merged in any way. They are rather similar (the resolveMenuItem code)! image It is probably used to feed the dropdown at the backend->sitemap, but something feels off with this combined with the translatable sitemap?

If it's your plugin, you can add the alternateLinks array directly in the original `pages.menuitem.resolveItem` event handler. The `winter.sitemap.processMenuItem` event is when extending a plugin that already exist to override its original menuItems.

**Comment by mjauvin** [comment]
> TESTING THE CHANGES I added these 5 file changes and 1 sitemap Definition.php change,
> 
> 5. Wondering if crawlers can crawl my sitemap on staging environment with Debug mode enabled (like SiteGuru or ScreamingFrog or...)

It's been changed now to only do the hack if the "?preview" query string is present in the request (see latest update to the Winter.Sitemap plugin PR)
 
> EDIT: I didn't manage to add my own Blog plugin. I am really close but am struggling with the component and slugs translations.
> 
> 1. I had to change some code, because otherwise it wouldn't pick up my component. For some reason it also needs the entire path
> 2. The only remaining issue is that the record is not being translated. Meaning that the slug isn't translated either.
> 
> For the rest it seems to work just fine!

I'd have to see more of your resolveMenuItem() method to help.

Also, is your custom Blog translatable?

**Comment by mjauvin** [comment]
It's not so much that you need anything special for the `page.menuitem.resolveItem` event handler, more so that if you use the `winter.sitemap.processMenuItem` event, the data needs to be sent with an extra layer into the result array.

**Comment by AIC-BV** [comment]
> It's not so much that you need anything special for the `page.menuitem.resolveItem` event handler, more so that if you use the `winter.sitemap.processMenuItem` event, the data needs to be sent with an extra layer into the result array.

But it should perhaps be mentionned in the documentation

**Comment by mjauvin** [comment]
> > It's not so much that you need anything special for the `page.menuitem.resolveItem` event handler, more so that if you use the `winter.sitemap.processMenuItem` event, the data needs to be sent with an extra layer into the result array.
> 
> But it should perhaps be mentioned in documentation, to prevent people from having that same issue as me in discord Because it is not like the blog plugin like mentioned at the [wn-pages-plugin](https://github.com/wintercms/wn-pages-plugin/blob/main/docs/documentation.md#creating-new-menu-item-types)

PRs welcome!

File: Plugin.php [hunk 5]
```diff
@@ -330,6 +336,64 @@ protected function extendWinterPagesPlugin(): void
...
+
+        Event::listen('winter.sitemap.makeUrlSet', function ($definition, $xml, $urlSet) {
+            if (Request::has('preview')) {
+                // hack to force browser to properly render the XML sitemap
+                $nsUrl = 'xmlns:xhtml-namespace-definition-URL-here';
+            } else {
+                $nsUrl = 'http://www.w3.org/1999/xhtml';
+            }
+            $urlSet->setAttribute('xmlns:xhtml', $nsUrl);
+        });
```
Comment (LukeTowers) [inline Plugin.php:380]: This should be removed and the sitemap plugin should instead implement sitemap stylesheets, which will fix the rendering issues in the browser when the data from the multilanguage sitemaps is present.

**Review by LukeTowers — CHANGES_REQUESTED** [review]

File: Plugin.php [hunk 6]
```diff
@@ -330,6 +336,58 @@ protected function extendWinterPagesPlugin(): void
...
+                if (in_array($item->type, $types)) {
+                    return $class::resolveMenuItem($item, $url, $theme);
+                }
+            }
+
+            return false;
+        });
+
+        Event::listen('winter.sitemap.makeUrlSet', function ($definition, $xml, $urlSet) {
+            $urlSet->setAttribute('xmlns:xhtml', 'http://www.w3.org/1999/xhtml');
```
Comment (LukeTowers) [inline Plugin.php:373]: Why is this required? Could this not just be the default?

Comment (mjauvin) [reply Plugin.php:374]: Because it's only required when we have xhtml:link nodes (which only the multilang requires for now).

Comment (LukeTowers) [reply Plugin.php:374]: Is there any reason to not include it when we don't? Any downside to it?

Comment (mjauvin) [reply Plugin.php:374]: Not that I know of, other than principle.

Comment (LukeTowers) [reply Plugin.php:374]: Let's just set it as the default in the sitemap plugin then

Comment (mjauvin) [reply Plugin.php]: done

**Review by LukeTowers — APPROVED** [review]
